src/common/client_response.hpp

```diff
@@ -0,0 +1,128 @@
+}
+//use this when creating an event
+//encodes enum to position in Enem
+template <class T>
+constexpr uint8_t GetPhaseIndex(T phase) {
+    return static_cast<size_t>(phase) - static_cast<size_t>(T::_first);
+}
+
+template <class T>
+constexpr T GetEnumFromPhaseIndex(size_t index) {
+    return static_cast<T>(static_cast<size_t>(T::_first) + index);
+}
+
+//define enum classes for responses here
+//and YES phase can have 0 responses
+//every enum must have "_first" and "_last"
+//"_first" ==  "previous_enum::_last" + 1
+//EVERY response shall have unique ID (so every button in GIU is unique)
+enum class PhasesLoadUnload : uint16_t {
+    _first = 0,
+    Parking,
+    WaitingTemp,
+    PreparingToRam,
+    Ramming,
+    Unloading,
+    Unloading2,
+    UserPush,
+    NozzleTimeout,
+    MakeSureInserted,
+    Inserting,
+    Loading,
+    Purging,
+    Purging2,
+    IsColor,
+    Unparking,
+    _last = Unparking
+};
+
+enum class PhasesTest : uint16_t {
+    _first = static_cast<uint16_t>(PhasesLoadUnload::_last) + 1,
+    Test1,
+    Test2,
+    _last = Test2
+};
+
+//static class for work with fsm responses (like button click)
+//encode responses - get them from marlin client, to marlin server and decode them again
+class ClientResponses {
+    ClientResponses() = delete;
+    ClientResponses(ClientResponses &) = delete;
+
+    //declare 2d arrays of single buttons for radio buttons
+    static const PhaseResponses LoadUnloadResponses[CountPhases<PhasesLoadUnload>()];
+    static const PhaseResponses TestResponses[CountPhases<PhasesTest>()];
+
+    //methods to "bind" button array with enum type
+    static const PhaseResponses &getResponsesInPhase(PhasesLoadUnload phase) { return LoadUnloadResponses[static_cast<size_t>(phase)]; }
+    static const PhaseResponses &getResponsesInPhase(PhasesTest phase) { return TestResponses[static_cast<size_t>(phase)]; }
+
+protected:
+    //get index of single response in PhaseResponses
+    template <class T>
+    static uint8_t GetIndex(T phase, Response response) {
+        const PhaseResponses &cmds = getResponsesInPhase(phase);
+        for (size_t i = 0; i < MAX_RESPONSES; ++i) {
+            if (cmds[i] == response)
+                return i;
+        }
+        return -1;
+    }
+
+    //get response from PhaseResponses by index
+    template <class T>
+    static Response GetResponse(T phase, uint8_t index) {
+        if (index > MAX_RESPONSES)
+            return Response::_none;
+        const PhaseResponses &cmds = getResponsesInPhase(phase);
+        return cmds[index];
+    }
+
+public:
+    //get all responses accepted in phase
+    template <class T>
+    static const PhaseResponses &GetResponses(T phase) {
+        return getResponsesInPhase(phase);
+    }
+
+    //encode radio button and clicked index into int
+    //use on client side
+    template <class T>
+    static uint32_t Encode(T phase, Response response) {
```

Comment:
Return value is of unsigned type, but we are returning negative number.

Reply:
fixed